src/sqlfluff/utils/reflow/depthmap.py

```diff
@@ -0,0 +1,128 @@
+"""The DepthMap class which is an enriched sequence of raw segments."""
+
+import logging
+from dataclasses import dataclass
+from typing import List, Sequence, Set, Tuple
+
+from sqlfluff.core.parser import BaseSegment
+from sqlfluff.core.parser.segments.raw import RawSegment
+
+
+reflow_logger = logging.getLogger("sqlfluff.rules.reflow")
+
+
+@dataclass(frozen=True)
+class DepthInfo:
+    """An object to hold the depth information for a specific raw segment."""
+
+    stack_depth: int
+    stack_hashes: List[int]
+    # This is a convenience cache to speed up operations.
+    stack_hash_set: Set[int]
```

Comment:
I notice `DepthMap` stores segment UUIDs, but here we're storing segment object hashes. Is there a reason for mixing approaches?

Reply:
This is an awkward corner of `Segment` comparison, and I don't really like it - but it felt like too much to change it as part of this PR.

`RawSegment` classes (and their decendants) have uuids for comparison.
`BaseSegment` classes (and their decendants) have hashes for comparison.

The reasons behind that are quite convoluted (and maybe need a rethink at some point).

Reply:
Thanks for the explanation! As I read farther in the PR, I saw you had a comment or two relating to this. Definitely not a blocker, just wanted to understand better. Perhaps someday, every segment could have a UUID. Anyway, not in scope for this work!